File: backend/src/main/java/com/mynewsblog/backend/model/Post.java
package com.mynewsblog.backend.model;

import jakarta.persistence.*;
import lombok.*;

import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;

@Entity
@Table(name = "posts")
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
@EqualsAndHashCode(onlyExplicitlyIncluded = true)
public class Post {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @EqualsAndHashCode.Include  // Use ID for equality checks
    private Long id;

    @Column(nullable = false, length = 200)
    private String title;

    @Column(nullable = false, unique = true, length = 255)
    private String slug;

    @Lob
    @Column(nullable = false, columnDefinition = "TEXT")
    private String content;

    // The main or featured image (optional)
    private String coverImage;

    // Which user created this post
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "user_id", nullable = false)
    private User author;

    // Which category the post belongs to
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "category_id", nullable = false)
    private Category category;

    private LocalDateTime createdAt;
    private LocalDateTime updatedAt;

    // One post can have many images
    @OneToMany(mappedBy = "post", cascade = CascadeType.ALL, orphanRemoval = true, fetch = FetchType.LAZY)
    @Builder.Default
    private List<PostImage> images = new ArrayList<>();

    @Column(nullable = false, columnDefinition = "bigint default 0")
    @Builder.Default
    private Long viewCount = 0L;

    @PrePersist
    public void onCreate() {
        createdAt = LocalDateTime.now();
        if (viewCount == null) {
            viewCount = 0L;
        }
    }

    @PreUpdate
    public void onUpdate() {
        updatedAt = LocalDateTime.now();
    }
}

File: frontend/src/app/models/post.ts
export interface PostResponse {
  id: number;
  title: string;
  content: string;
  coverImage?: string | null;
  categoryId?: number;
  categoryName: string;
  authorUsername: string;
  createdAt: string;
  updatedAt?: string | null;
  viewCount?: number;
  imageUrls: string[];
}

// request payloads expected by backend DTOs
export interface CreatePostRequest {
  title: string;
  content: string;
  categoryId: number;
  coverImage?: string | null;
  imageUrls?: string[];
}

export interface UpdatePostRequest {
  title?: string;
  content?: string;
  categoryId?: number;
  coverImage?: string | null;
  imageUrls?: string[];
}
